File: plugins/plugin-auto-trader/rust/src/actions/execute_live_trade.rs
use async_trait::async_trait;
use serde_json::{json, Value};

use crate::service::TradingService;
use crate::types::TradeDirection;
use crate::{Action, ActionExample, ActionResult};

pub struct ExecuteLiveTradeAction;

#[async_trait]
impl Action for ExecuteLiveTradeAction {
    fn name(&self) -> &str {
        "EXECUTE_LIVE_TRADE"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "EXECUTE_TRADE",
            "PLACE_TRADE",
            "MAKE_TRADE",
            "BUY_TOKEN",
            "SELL_TOKEN",
        ]
    }

    fn description(&self) -> &str {
        "Execute a live trade (buy/sell) on the mock exchange."
    }

    async fn validate(&self, _message: &Value, _state: &Value) -> bool {
        true
    }

    async fn handler(
        &self,
        message: &Value,
        state: &Value,
        service: Option<&TradingService>,
    ) -> ActionResult {
        let Some(svc) = service else {
            return ActionResult {
                success: false,
                text: "TradingService is not available.".to_string(),
                data: None,
                error: Some("missing_service".to_string()),
            };
        };

        let token = state
            .get("token")
            .and_then(|v| v.as_str())
            .unwrap_or("SOL");

        let amount = state
            .get("amount")
            .and_then(|v| v.as_f64())
            .unwrap_or(1.0);

        let text = message
            .pointer("/content/text")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let direction_str = state
            .get("direction")
            .and_then(|v| v.as_str())
            .unwrap_or(text);

        let direction = if direction_str.to_lowercase().contains("sell") {
            TradeDirection::Sell
        } else {
            TradeDirection::Buy
        };

        match svc.execute_trade(token, direction, amount).await {
            Ok(trade) => {
                let text = format!(
                    "Trade executed: {} {:.4} {} @ ${:.4} (ID: {})",
                    trade.direction, trade.amount, trade.token, trade.price, trade.id
                );
                ActionResult {
                    success: true,
                    text,
                    data: Some(json!(trade)),
                    error: None,
                }
            }
            Err(e) => ActionResult {
                success: false,
                text: e,
                data: None,
                error: Some("trade_failed".to_string()),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![
            ActionExample {
                user_message: "Buy 10 SOL".to_string(),
                agent_response: "Trade executed: BUY 10.0000 SOL @ $150.00".to_string(),
            },
            ActionExample {
                user_message: "Sell 5 ETH".to_string(),
                agent_response: "Trade executed: SELL 5.0000 ETH @ $2500.00".to_string(),
            },
        ]
    }
}

File: plugins/plugin-auto-trader/python/elizaos_plugin_auto_trader/actions/execute_live_trade.py
from __future__ import annotations

from dataclasses import dataclass

from elizaos_plugin_auto_trader.actions.common import ActionResult, Message
from elizaos_plugin_auto_trader.service import TradingService


@dataclass
class ExecuteLiveTradeAction:
    @property
    def name(self) -> str:
        return "EXECUTE_LIVE_TRADE"

    async def validate(self, _message: Message, _state: dict) -> bool:  # type: ignore[type-arg]
        return True

    async def handler(
        self,
        message: Message,
        state: dict,  # type: ignore[type-arg]
        service: TradingService | None = None,
    ) -> ActionResult:
        if service is None:
            return ActionResult(False, "TradingService is not available.", "missing_service")

        token = state.get("token", "SOL")
        amount = float(state.get("amount", 1.0))
        text = (message.get("content") or {}).get("text") or ""
        direction_hint = state.get("direction", text)
        direction = "SELL" if "sell" in str(direction_hint).lower() else "BUY"

        try:
            trade = await service.execute_trade(token, direction, amount)
        except RuntimeError as exc:
            return ActionResult(False, str(exc), "trade_failed")

        return ActionResult(
            True,
            f"Trade executed: {trade.direction} {trade.amount:.4f} "
            f"{trade.token} @ ${trade.price:.4f} (ID: {trade.id})",
            data=trade,
        )
